samples/nrf_desktop/src/modules/dfu.c

```diff
@@ -0,0 +1,242 @@
+
+static const struct flash_area *flash_area;
+
+
+static void set_ble_latency(bool dfu_start)
+{
+	if (!active_conn) {
+		LOG_INF("No active_connection");
+		return;
+	}
+
+	struct bt_conn_info info;
+
+	int err = bt_conn_get_info(active_conn, &info);
+	if (err) {
+		LOG_WRN("Cannot get conn info (%d)", err);
+		return;
+	}
+
+	if (IS_ENABLED(CONFIG_BT_PERIPHERAL) &&
+	    (info.role == BT_CONN_ROLE_SLAVE)) {
+		const struct bt_le_conn_param param = {
+			.interval_min = info.le.interval,
+			.interval_max = info.le.interval,
+			.latency = (dfu_start) ? (0) : (CONFIG_BT_PERIPHERAL_PREF_SLAVE_LATENCY),
+			.timeout = info.le.timeout
+		};
+
+		err = bt_conn_le_param_update(active_conn, &param);
+		if (err) {
+			LOG_WRN("Cannot update parameters (%d)", err);
+			return;
+		}
+	}
+
+	LOG_INF("BLE latency %screased for DFU", dfu_start ? "de" : "in");
+}
+
+static void dfu_timeout_handler(struct k_work *work)
+{
+	LOG_WRN("DFU timed out");
+
+	set_ble_latency(false);
+
+	if (flash_area) {
+		flash_area_close(flash_area);
+		flash_area = NULL;
+	}
+}
+
+static void handle_dfu_data(u8_t *data, size_t size)
+{
+	int err;
+
+	u32_t img_length;
+	u32_t chunk_offset;
+	u32_t chunk_size;
+
+	size_t min_size = sizeof(img_length) + sizeof(chunk_offset);
+
+	if (size < min_size) {
+		LOG_INF("Invalid DFU data header");
+		return;
+	}
+
+	img_length = sys_get_le32(&data[0]);
+	chunk_offset = sys_get_le32(&data[sizeof(img_length)]);
+	chunk_size = size - min_size;
+
+	LOG_INF("DFU data received img_length:%" PRIu32
+		" chunk_offset:%" PRIu32 " chunk_size:%" PRIu32,
+		img_length, chunk_offset, chunk_size);
+
+	static size_t cur_offset;
+
+	if (chunk_offset == 0) {
+		set_ble_latency(true);
+
+		if (cur_offset) {
+			LOG_WRN("Previous DFU operation interrupted");
+		}
+
+		err = flash_area_open(DT_FLASH_AREA_IMAGE_1_ID, &flash_area);
+		if (err) {
+			LOG_ERR("Cannot open flash area (%d)", err);
+			goto error;
+		}
+
+		cur_offset = 0;
+	}
+
+	if (!flash_area) {
+		LOG_WRN("DFU was not started");
+		goto error;
+	}
+
+	if (flash_area->fa_size < img_length) {
+		LOG_WRN("Insufficient space for DFU");
+		goto error;
+	}
+
+	if (chunk_offset != cur_offset) {
+		LOG_WRN("Invalid chunk sequence");
+		goto error;
+	}
+
+	if (!chunk_size) {
+		LOG_WRN("No data to store");
+		goto error;
+	}
+
+	if (cur_offset % FLASH_PAGE_SIZE == 0) {
+		err = flash_area_erase(flash_area, cur_offset, FLASH_PAGE_SIZE);
+		if (err) {
+			LOG_ERR("Cannot erase page (%d)", err);
+			goto error;
+		}
+	}
+
+	void *chunk_data = &data[min_size];
+
+	err = flash_area_write(flash_area, cur_offset, chunk_data, chunk_size);
+	if (err) {
+		LOG_ERR("Cannot write data (%d)", err);
+		goto error;
+	}
+
+	cur_offset += chunk_size;
+
+	LOG_INF("DFU chunk written");
+	if (img_length == cur_offset) {
+		LOG_INF("DFU image written");
+		boot_request_upgrade(false);
+
+		goto error;
+	} else {
+		k_delayed_work_submit(&dfu_timeout, DFU_TIMEOUT);
+	}
+
+	return;
+
+error:
```

Comment:
maybe we could call it dfu_finish, or something like that, since we do it also after proper DFU image write?

Reply:
Sure, fixed